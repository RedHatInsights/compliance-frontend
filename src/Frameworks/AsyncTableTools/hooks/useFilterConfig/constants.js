export const TABLE_STATE_NAMESPACE = 'filters';
