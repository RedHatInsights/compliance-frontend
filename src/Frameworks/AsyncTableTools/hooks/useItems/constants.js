export const TABLE_STATE_NAMESPACE = 'items';
