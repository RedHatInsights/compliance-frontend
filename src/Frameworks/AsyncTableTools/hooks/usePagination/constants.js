export const TABLE_STATE_NAMESPACE = 'pagination';
