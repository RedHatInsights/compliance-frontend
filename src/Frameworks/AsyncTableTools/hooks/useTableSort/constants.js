export const TABLE_STATE_NAMESPACE = 'sort';
