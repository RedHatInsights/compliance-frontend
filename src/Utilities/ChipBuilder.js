import { conditionalFilterType } from '@redhat-cloud-services/frontend-components/ConditionalFilter';

const findWithString = (value) => (
    (item) => (
        String(item.value) === String(value)
    )
);

class ChipBuilder {
    constructor(filterConfig) {
        this.filterConfig = filterConfig;
        this.config = this.filterConfig.config;
    }

    textFilterChip = (category, currentValue) => (currentValue && currentValue !== '' ? {
        category,
        chips: [{ name: currentValue }]
    } : null)

    checkboxFilterChip = (category, currentValue) => (currentValue && currentValue.length > 0 ? {
        category,
        chips: currentValue.map((value) => (
            { name: this.filterConfig.labelForValue(value, category) }
        ))
    } : null)

    radioFilterChip = (category, currentValue) => (currentValue && currentValue !== '' ? {
        category,
        chips: [{ name: this.filterConfig.labelForValue(currentValue, category) }]
    } : null)

    groupFilterChip = (category, currentValue = {}, categoryConfig) => {
        const items = Object.keys(currentValue).flatMap((groupValue) => {
            const group = categoryConfig.items.find(findWithString(groupValue));

            return Object.keys(currentValue[groupValue]).map((itemValue) => (
                currentValue[groupValue][itemValue] ? group?.items.find(findWithString(itemValue)) : null
            ));
        }).filter((v) => (!!v));

        return currentValue && items.length > 0 ? {
            category,
            chips: items.map((item) => ({ name: item.label }))
        } : null;
    }

    chipFor = (filter, currentValue) => {
        const categoryConfig = this.filterConfig.getCategoryForLabel(filter);
        const { label, type } = categoryConfig ? categoryConfig : { label: filter, type: null };

        switch (type) {
            case conditionalFilterType.text:
                return this.textFilterChip(label, currentValue);

            case conditionalFilterType.checkbox:
                return this.checkboxFilterChip(label, currentValue);

            case conditionalFilterType.radio:
                return this.radioFilterChip(label, currentValue);

            case conditionalFilterType.group:
                return this.groupFilterChip(label, currentValue, categoryConfig);

            default:
                return null;
        }
    }

    chipsFor = (filters) => (
        Object.keys(filters).map((filter) => (
            this.chipFor(filter, filters[filter])
        )).filter((f) => (!!f))
    )
}

export default ChipBuilder;
