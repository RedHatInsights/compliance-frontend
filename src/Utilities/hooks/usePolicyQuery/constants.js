import gql from 'graphql-tag';

export const POLICY_QUERY = gql`
  query PQ_Profile($policyId: String!) {
    profile(id: $policyId) {
      id
      name
      refId
      external
      description
      totalHostCount
      compliantHostCount
      complianceThreshold
      osMajorVersion
      supportedOsVersions
      lastScanned
      policyType
      policy {
        id
        name
        refId
        profiles {
          id
          name
          refId
          osMinorVersion
          osMajorVersion
          parentProfileId
          benchmark {
            id
            title
            latestSupportedOsMinorVersions
            osMajorVersion
            version
            profiles {
              id
              refId
              ssgVersion
            }
          }
          rules {
            id
            title
            severity
            rationale
            refId
            description
            remediationAvailable
            references
            identifier
            precedence
            values
          }
        }
      }
      businessObjective {
        id
        title
      }
      hosts {
        id
        osMinorVersion
        osMajorVersion
      }
    }
  }
`;

export const POLICY_RULE_TREES_QUERY = gql`
  query PQ_Profile($policyId: String!) {
    profile(id: $policyId) {
      policy {
        profiles {
          id
          benchmark {
            ruleTree
          }
        }
      }
    }
  }
`;

export const POLICY_VALUE_DEFINITONS_QUERY = gql`
  query PQ_Profile($policyId: String!) {
    profile(id: $policyId) {
      policy {
        profiles {
          id
          values
          benchmark {
            valueDefinitions {
              id
              refId
              title
              valueType
              defaultValue
              description
            }
          }
        }
      }
    }
  }
`;
