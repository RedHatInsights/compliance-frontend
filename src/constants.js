import { version } from './../package.json';
import { conditionalFilterType } from '@redhat-cloud-services/frontend-components';

export const COMPLIANCE_API_ROOT = '/api/compliance';
export const COMPLIANCE_UI_ROOT = '/rhel/compliance';
export const COMPLIANCE_WS_ROOT = process.env.NODE_ENV === 'production'
    ? 'wss://localhost:3000/cable'
    : 'ws://localhost:3000/cable';
export const INVENTORY_API_ROOT = '/api/inventory/v1';

export const API_HEADERS = {
    'X-Insights-Compliance': version,
    'Content-Type': 'application/json',
    Accept: 'application/json'
};

export const DEFAULT_SYSTEMS_FILTER_CONFIGURATION = [
    {
        type: conditionalFilterType.text,
        label: 'Name',
        event: 'onSubmit',
        filterString: (value) => (`name ~ ${value}`)
    }
];

export const COMPLIANT_SYSTEMS_FILTER_CONFIGURATION = [
    {
        type: conditionalFilterType.checkbox,
        label: 'Compliant',
        filterString: (value) => (`compliant = ${value}`),
        items: [
            { label: 'Compliant', value: 'true' },
            { label: 'Non-compliant', value: 'false' }
        ]
    },
    {
        type: conditionalFilterType.checkbox,
        label: 'Compliance score',
        filterString: (value) => {
            const scoreRange = value.split('-');
            return `compliance_score >= ${scoreRange[0]} and compliance_score <= ${scoreRange[1]}`;
        },
        items: [
            { label: '90 - 100%', value: '90-100' },
            { label: '70 - 89%', value: '70-89' },
            { label: '50 - 69%', value: '50-69' },
            { label: 'Less than 50%', value: '0-49' }
        ]
    }
];
