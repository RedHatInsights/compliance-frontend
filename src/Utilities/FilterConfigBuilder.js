import {
    conditionalFilterType
} from '@redhat-cloud-services/frontend-components';
import { stringToId } from 'Utilities/TextHelper';

export class FilterConfigBuilder {
    constructor(config) {
        this.config = config;
    }

    toTextFilterConfig = (item, handler, value) => ({
        type: conditionalFilterType.text,
        label: item.label,
        id: stringToId(item.label),
        filterValues: {
            value,
            onSubmit: (_event, selectedValues) => {
                handler(stringToId(item.label), selectedValues);
            }
        }
    });

    toCheckboxFilterConfig = (item, handler, value) => ({
        type: conditionalFilterType.checkbox,
        label: item.label,
        id: stringToId(item.label),
        filterValues: {
            value,
            items: item.items,
            onChange: (_event, selectedValues) => {
                handler(stringToId(item.label), selectedValues);
            }
        }
    });

    toFilterConfigItem = (item, handler, value) => {
        if (item.type === conditionalFilterType.text) {
            return this.toTextFilterConfig(item, handler, value);
        } else if (item.type === conditionalFilterType.checkbox) {
            return this.toCheckboxFilterConfig(item, handler, value);
        }
    };

    buildConfiguration = (handler, states, props = {}) => ({
        ...props,
        items: this.config.map((item) => (
            this.toFilterConfigItem(item, handler, states[stringToId(item.label)])
        ))
    });

    initialDefaultState = () => {
        let initState = {};
        this.config.forEach((filter) => {
            const filterStateName = filter.label.replace(' ', '').toLowerCase();
            if (filter.type ===  conditionalFilterType.checkbox) {
                initState[filterStateName] = [];
            } else if (filter.type === conditionalFilterType.text) {
                initState[filterStateName] = '';
            } else {
                return;
            }
        });
        return initState;
    };

    categoryLabelForValue = (value) => {
        const category = this.config.filter((category) => {
            return category.items ?
                category.items.map((item) => item.value).includes(value) : false;
        });

        return category ? category[0].label : value;
    };

    labelForValue = (value, category) => {
        try {
            return this.config.filter((item) => (item.label === category))[0]
            .items.filter((item) => (item.value === value))[0].label;
        }
        catch (_) {
            // eslint-disable-next-line no-console
            console.info('No label found for ' + value + ' in ', category);
            return value;
        }
    };

    valueForLabel = (label, category) => {
        try {
            return this.config.filter((item) => (item.label === category))[0]
            .items.filter((item) => (item.label === label))[0].value;
        }
        catch (_) {
            // eslint-disable-next-line no-console
            console.info('No value found for ' + label + ' in ' + category);
            return label;
        }
    };
}
