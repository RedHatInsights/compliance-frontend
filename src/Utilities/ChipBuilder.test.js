import { FilterConfigBuilder } from './FilterConfigBuilder';
import { FILTER_CONFIGURATION } from '../constants';

import { ChipBuilder } from './ChipBuilder';

describe('ChipBuilder#getChipsFor', () => {
    const builder = new FilterConfigBuilder(FILTER_CONFIGURATION);
    let chipBuilder;

    beforeEach(() => {
        chipBuilder = new ChipBuilder(builder);
    });

    it('returns a filterConfig', () => {
        chipBuilder.chipsFor({
            compliant: ['false'],
            name: 'Text search',
            compliancescore: ['50-69', '70-89'],
            invalidFilter: ''
        }).then((results) => {
            expect(results).toMatchSnapshot();
        });
    });
});
