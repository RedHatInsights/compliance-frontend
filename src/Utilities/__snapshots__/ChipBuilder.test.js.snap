// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChipBuilder#getChipsFor returns a filterConfig 1`] = `
Array [
  Object {
    "category": "Compliant",
    "chips": Array [
      Object {
        "name": "Non-compliant",
      },
    ],
  },
  Object {
    "category": "Name",
    "chips": Array [
      Object {
        "name": "Text search",
      },
    ],
  },
  Object {
    "category": "Compliance score",
    "chips": Array [
      Object {
        "name": "50 - 69%",
      },
      Object {
        "name": "70 - 89%",
      },
    ],
  },
]
`;
